### PR TITLE
zenergy: init at 0-unstable-2024-10-10

### PR DESCRIPTION
**Pull request by wizardlink**
## Description of changes

Adds a new kernel module, [`zenergy`](https://github.com/BoukeHaarsma23/zenergy), this is necessary for accessing CPU power draw data without root permission, specifically [MangoHud](https://github.com/flightlessmango/MangoHud) requires it to be able to show the user their CPU power draw.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by wizardlink** [comment]
I apologize if I missed something in advance, first time contributing to the repository! I am a little confused though with the testing and release node markdown, for the testing I couldn't find an example of a kernel module being tested, if it's even possible with the suite available; whereas the markdown file, it implies that not all additions will be covered, so would this PR be applicable still? Not quite sure.

File: pkgs/os-specific/linux/zenergy/default.nix [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+  outputs = [ "out" ];
+
+  meta = with lib; {
+    description = "Based on AMD_ENERGY driver, but with some jiffies added so non-root users can read it safely.";
+    homepage = "https://github.com/BoukeHaarsma23/zenergy";
+    license = licenses.gpl2Only;
+    maintainers = with maintainers; [ wizardlink ];
+    platforms = platforms.linux;
+    outputsToInstall = [ "out" ];
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/zenergy/default.nix:42]:
```suggestion
```
That's also the default

File: pkgs/os-specific/linux/zenergy/default.nix [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+  '';
+
+  nativeBuildInputs = [ kmod ] ++ kernel.moduleBuildDependencies;
+
+  installPhase = ''
+    make modules_install KDIR=${kernelDirectory}
+  '';
+
+  outputs = [ "out" ];
+
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/zenergy/default.nix:35]:
```suggestion
```
That's the default

File: pkgs/os-specific/linux/zenergy/default.nix [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  preBuild = ''
+    substituteInPlace Makefile --replace-fail "PWD modules_install" "PWD INSTALL_MOD_PATH=$out modules_install"
+  '';
+
+  nativeBuildInputs = [ kmod ] ++ kernel.moduleBuildDependencies;
+
+  installPhase = ''
+    make modules_install KDIR=${kernelDirectory}
+  '';
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/zenergy/default.nix:32]: It should be enough to set makeTarget and makeFlags like already done

File: pkgs/os-specific/linux/zenergy/default.nix [hunk 4]
```diff
@@ -0,0 +1,44 @@
+{ lib, stdenv, fetchFromGitHub, kernel, kmod }:
+
+let
+  kernelDirectory = "${kernel.dev}/lib/modules/${kernel.modDirVersion}/build";
+  version = "a3e124477ee8197015481156b90100d49fa3cd84";
+in
+stdenv.mkDerivation {
+  inherit version;
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/zenergy/default.nix:8]: Please use the unstable version format. See contributing guide for details.

**Comment by SuperSandro2000** [comment]
> I am a little confused though with the testing and release note markdown, for the testing I couldn't find an example of a kernel module being tested, if it's even possible with the suite available; whereas the markdown file, it implies that not all additions will be covered, so would this PR be applicable still? Not quite sure.

If you are doing the change on top of nixos-unstable, you can just rebuild your system with it and try to use the kernel module like normally done.
We don't need a release note entry here.

**Comment by wizardlink** [comment]
I've been using that kernel module locally since the PR has been open, so I can attest that it's in working condition!

File: pkgs/os-specific/linux/zenergy/default.nix (hunk 3 above)
Comment (wizardlink) [reply pkgs/os-specific/linux/zenergy/default.nix:31]: I searched in the repository for examples and tried to replicate, but in the `installPhase` the `KDIR` environment wasn't set on top of not running `modules_install` with `makeTargets = [ "modules_install" ]`.

Am I missing something?

Comment (wizardlink) [reply pkgs/os-specific/linux/zenergy/default.nix:31]: @SuperSandro2000 can you also give me an update on this?

Comment (Artturin) [reply pkgs/os-specific/linux/zenergy/default.nix:31]: `  installTargets = [ "modules_install" ];` works

Comment (wizardlink) [reply pkgs/os-specific/linux/zenergy/default.nix]: Thank you for the heads-up! Can't believe I hadn't noticed that in the docs.

**Pull request merged**